src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -78,10 +79,10 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
                 descriptors = matchingDescriptors.Concat(descriptors);
             }
 
-            var applicableDescriptors = ApplyRequiredAttributes(descriptors, attributeNames);
+            var applicableDescriptors = ApplyRequiredAttributes(descriptors, attributes);
             applicableDescriptors = ApplyParentTagFilter(applicableDescriptors, parentTagName);
 
-            return applicableDescriptors;
+            return applicableDescriptors.ToArray();
```

Comment:
I was debugging this and found that we re-iterate this enumerable like 3x per descriptor found. Made debugging incredibly difficult.